Framework/Kernel/src/Matrix.cpp

```diff
@@ -988,37 +990,51 @@ yes invert the matrix using analytic formula. If not then use standard Invert
     // use analytic expression as described in G Y Hu and R F O'Connell (1996)
     T scalefactor = numRows() > 1 ? m_rawData[1][0] : 1;
     *this /= scalefactor;
-    T D = m_rawData[0][0];
-    auto k = static_cast<T>(numRows());
+    long double D = m_rawData[0][0];
+    long double k = static_cast<long double>(numRows());
+    if ((std::abs(D) >= 2.0) && ((k * std::acosh(D / 2.0)) > std::acosh(std::numeric_limits<long double>::max()))) {
+      *this *= scalefactor;
+      Invert();
+      return;
+    }
+
     for (size_t i = 0; i < numRows(); i++) {
       for (size_t j = 0; j < numCols(); j++) {
-        T lambda;
-        auto iMinusj = static_cast<T>(i) - static_cast<T>(j);
-        auto iPlusj = static_cast<T>(i) + static_cast<T>(j);
+        long double lambda;
+        long double iMinusj;
+        if (i > j)
+          iMinusj = static_cast<long double>(i - j);
+        else
+          iMinusj = static_cast<long double>(j - i);
+        long double iPlusj = static_cast<long double>(i + j);
```

Comment:
This is a nitpick, but I guess `iPlusj` could still be `auto`, as the type should be defined by the type in  the `static_cast`.